modules/transactions/src/main/java/org/apache/ignite/internal/tx/TxMeta.java

```diff
@@ -68,6 +74,21 @@ public Collection<TablePartitionId> enlistedPartitions() {
         return commitTimestamp;
     }
 
+    @Override
+    public TxMetaMessage toTransactionMetaMessage(ReplicaMessagesFactory replicaMessagesFactory, TxMessagesFactory txMessagesFactory) {
+        var enlistedPartitionMessages = new ArrayList<TablePartitionIdMessage>(enlistedPartitions.size());
+
+        for (TablePartitionId enlistedPartition : enlistedPartitions) {
+            enlistedPartitionMessages.add(toTablePartitionIdMessage(replicaMessagesFactory, enlistedPartition));
+        }
+
+        return txMessagesFactory.txMetaMessage()
+                .txStateInt(txState.ordinal())
+                .commitTimestampLong(hybridTimestampToLong(commitTimestamp))
+                .enlistedPartitions(enlistedPartitionMessages)
```

Comment:
Noticed that in messages constructors we don't create defensive copies for collections, maybe we should add a TODO to fix that?

Reply:
I saw that now, when we deserialize, collections also do not become safe from changes. I think this can be solved as a separate problem, without TODO.